Documentation/operations/upgrade.rst

```diff
@@ -339,14 +339,15 @@ Annotations:
 * When using the ENI-based IPAM in conjunction with the ``--eni-tags``, failures
   to create tags are treated as errors which will result in ENIs not being
   created. Ensure that the ``ec2:CreateTags`` IAM permissions are granted.
-* To prevent Pods being scheduled by other CNI plugins during Cilium agent
-  downtime (upgrades, crashes,..), Cilium now assumes ownership of the node's
-  CNI configuration directory (``/etc/cni/net.d``) by default.
-  Existing CNI configurations are renamed to ``*.cilium_bak``, which will
-  disable solutions like [CNI-Genie](https://github.com/cni-genie/CNI-Genie)
-  as well as the default CNI plugins shipped by most managed Kubernetes
-  distributions. Set the ``cni.exclusive=false`` Helm flag to disable this
-  behaviour.
+* Cilium now takes ownership of the ``/etc/cni/net.d/`` directory on the host
+  by default. During agent startup, Cilium replaces all CNI configs containing
+  the word ``cilium``, and non-Cilium CNI configurations are renamed to
+  ``*.cilium_bak``. During agent shutdown, all Cilium CNI configs are removed.
```

Comment:
Questions for clarification, not specifically feedback on the upgrade note:

* Does Cilium CNI config removal happen during regular Cilium restart and/or Cilium upgrade? (That sounds like it might regress functionality, though maybe k8s retries will work in a similar way to today)
* Are the `*.cilium_bak` files restored to original paths during agent shutdown (and/or restart / upgrade)?

Reply:
> Does Cilium CNI config removal happen during regular Cilium restart and/or Cilium upgrade?

Yep, it does. Both the CNI config and binary are installed and uninstalled during start and stop respectively. This addresses https://github.com/cilium/cilium/issues/13152.

> That sounds like it might regress functionality

Anything in particular you're worried about? Since k8s 1.16, dockershim health checks all CNI plugins every 5s and marks them as unavailable, causing them to be skipped, and the next-healthy CNI plugin used for scheduling containers. This poses a security risk during upgrades (as well as `cilium-cni` failing `CNI CHECK` calls at runtime), hence 'exclusive' mode was introduced. The only way to currently guarantee a pod gets scheduled by Cilium is by moving all other CNIs out of the way.

> Are the *.cilium_bak files restored to original paths during agent shutdown (and/or restart / upgrade)?

That's currently done by `cilium cleanup`.

Reply:
I think I'm just worried about how the different layers handle the backoffs in the case of missing config. The model I had in my mind of how it worked was that we should always force the configuration through the Cilium CNI and that the only kubelet mechanism in place was to call into the CNI. If Cilium is down, then cilium-cni would run and handle backoffs etc. However, based on the above description, I was missing the CNI plugin health check piece of this which introduces a different set of challenges. Based on that, I follow your reasoning and it sounds like the best option we have.

Just for context I'd be curious what the errors look like when deployment failure occurs during Cilium restart/outage, if you happen to have any on-hand it'd be nice to put them front and center in the PR description or something so that users might be able to find & reason about the error message (=> "oh, it's just because Cilium is restarting, pod deployment will work again soon"). But to satisfy my own curiosity, maybe I should just try deploying some pods during Cilium restart and observe the behaviour to gather this info..

Reply:
For what it's worth, the behaviour I observe locally in this scenario is that pods stay in "Pending" state and `kubectl describe` shows:

```
  Warning  FailedScheduling  2s (x4 over 104s)  default-scheduler  0/1 nodes are available: 1 node(s) had taint {node.kubernetes.io/not-ready: }, that the pod didn't tolerate.
```

So basically the node is "not-ready" until Cilium restarts.

Reply:
Yup, that's what I remember observing. This is exactly what we want. :+1: